### PR TITLE
Improve Vertex mlops blueprint

### PR DESCRIPTION
**Pull request by lcaggio**
in this PR we are improving the Vertex MLOps blueprint supporting:
 - Vertex Metastore
 - CMEK on all resources
 - User managed workbench
 - Consolidate variables

### CODE REVIEW

File: blueprints/data-solutions/vertex-mlops/main.tf [hunk 1]
```diff
@@ -190,19 +190,19 @@ module "cloudnat" {
 
 module "project" {
   source          = "../../../modules/project"
-  name            = var.project_id
-  parent          = try(var.project_create.parent, null)
-  billing_account = try(var.project_create.billing_account_id, null)
-  project_create  = var.project_create != null
+  name            = var.project_config.project_id
+  parent          = try(var.project_config.parent, null)
+  billing_account = try(var.project_config.billing_account_id, null)
```
Comment (juliocc) [inline blueprints/data-solutions/vertex-mlops/main.tf:195]: this `try()` is not needed and can hide other issues. Make the variable `nullable = false` and provide sensible defaults

File: blueprints/data-solutions/vertex-mlops/main.tf [hunk 2]
```diff
@@ -190,19 +190,19 @@ module "cloudnat" {
...
-    "roles/aiplatform.user"         = [module.service-account-mlops.iam_email]
+    "roles/aiplatform.user"         = [module.service-account-mlops.iam_email, module.service-account-notebook.iam_email]
     "roles/artifactregistry.reader" = [module.service-account-mlops.iam_email]
     "roles/artifactregistry.writer" = [module.service-account-github.iam_email]
-    "roles/bigquery.dataEditor"     = [module.service-account-mlops.iam_email]
-    "roles/bigquery.jobUser"        = [module.service-account-mlops.iam_email]
-    "roles/bigquery.user"           = [module.service-account-mlops.iam_email]
+    "roles/bigquery.dataEditor"     = [module.service-account-mlops.iam_email, module.service-account-notebook.iam_email]
+    "roles/bigquery.jobUser"        = [module.service-account-mlops.iam_email, module.service-account-notebook.iam_email]
+    "roles/bigquery.user"           = [module.service-account-mlops.iam_email, module.service-account-notebook.iam_email]
```
Comment (juliocc) [inline blueprints/data-solutions/vertex-mlops/main.tf:205]: split identities into multiple lines

File: blueprints/data-solutions/vertex-mlops/main.tf [hunk 3]
```diff
@@ -223,28 +225,40 @@ module "project" {
...
-    # }
-  }
-
   service_encryption_key_ids = {
+    aiplatform = [try(local.service_encryption_keys.aiplatform, null)]
     bq         = [try(local.service_encryption_keys.bq, null)]
-    compute    = [try(local.service_encryption_keys.compute, null)]
     cloudbuild = [try(local.service_encryption_keys.storage, null)]
-    notebooks  = [try(local.service_encryption_keys.compute, null)]
+    notebooks  = [try(local.service_encryption_keys.notebooks, null)]
```
Comment (juliocc) [inline blueprints/data-solutions/vertex-mlops/main.tf:238]: same comment as before regarding `try()`

File: blueprints/data-solutions/vertex-mlops/variables.tf [hunk 4]
```diff
@@ -135,18 +122,20 @@ variable "repo_name" {
...
-}
-
-variable "service_encryption_keys" { # service encription key
+variable "service_encryption_keys" {
   description = "Cloud KMS to use to encrypt different services. Key location should match service region."
   type = object({
-    bq      = string
-    compute = string
-    storage = string
+    aiplatform    = optional(string, null)
```
Comment (juliocc) [inline blueprints/data-solutions/vertex-mlops/variables.tf:128]: `optional(string)`

File: blueprints/data-solutions/vertex-mlops/variables.tf [hunk 5]
```diff
@@ -135,18 +122,20 @@ variable "repo_name" {
...
+    bq            = optional(string, null)
+    notebooks     = optional(string, null)
+    secretmanager = optional(string, null)
+    storage       = optional(string, null)
   })
-  default = null
-}
\ No newline at end of file
+  default = {
```
Comment (juliocc) [inline blueprints/data-solutions/vertex-mlops/variables.tf:134]: `default={}`

File: blueprints/data-solutions/vertex-mlops/vertex.tf [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+      }
+      data_disk {
+        initialize_params {
+          disk_size_gb = "100"
+          disk_type    = "PD_STANDARD"
+        }
+      }
+    }
+  }
+}
```
Comment (juliocc) [inline blueprints/data-solutions/vertex-mlops/vertex.tf:81]: all the hardcoded values in this resource should probably become their own variable

File: blueprints/data-solutions/vertex-mlops/vertex.tf [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+resource "google_notebooks_instance" "playground" {
+  for_each     = { for k, v in var.notebooks : k => v if v.type == "USER_MANAGED" }
+  name         = "${var.prefix}-${each.key}"
+  location     = format("%s-%s", var.region, "b")
```
Comment (juliocc) [inline blueprints/data-solutions/vertex-mlops/vertex.tf:86]: use interpoaltion

File: blueprints/data-solutions/vertex-mlops/vertex.tf [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+  container_image {
+    repository = "gcr.io/deeplearning-platform-release/base-cpu"
+    tag        = "latest"
+  }
+
+  install_gpu_driver = true
+  boot_disk_type     = "PD_SSD"
+  boot_disk_size_gb  = 110
+  disk_encryption    = try(local.service_encryption_keys.notebooks != null, false) ? "CMEK" : null
+  kms_key            = try(local.service_encryption_keys.notebooks, null)
```
Comment (juliocc) [inline blueprints/data-solutions/vertex-mlops/vertex.tf:99]: idem `try()`

**Review by juliocc — CHANGES_REQUESTED** [review]

File: blueprints/data-solutions/vertex-mlops/vertex.tf (hunk 6 above)
Comment (lcaggio) [reply blueprints/data-solutions/vertex-mlops/vertex.tf:81]: I will address this is a future PR.

File: blueprints/data-solutions/vertex-mlops/variables.tf (hunk 4 above)
Comment (lcaggio) [reply blueprints/data-solutions/vertex-mlops/variables.tf:128]: I have the default value to remove the try() in the code.

Comment (juliocc) [reply blueprints/data-solutions/vertex-mlops/variables.tf:128]: `optional(string, null)` is the same as `optional(string)`

> The optional modifier takes one or two arguments.
>
>Type: (Required) The first argument specifies the type of the attribute.
> Default: (Optional) The second argument defines the default value that Terraform should use if the attribute is not present. This must be compatible with the attribute type. If not specified, **Terraform uses a null value of the appropriate type as the default**.

File: blueprints/data-solutions/vertex-mlops/metadata.yaml [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+apiVersion: blueprints.cloud.google.com/v1alpha1
+kind: BlueprintMetadata
+metadata:
+  name: terraform-google-fabric-data_solutiont-vertex_mlops
```
Comment (juliocc) [inline blueprints/data-solutions/vertex-mlops/metadata.yaml:18]: typo ("solutiont") but I'd use just fabric-vertex-mlops

File: blueprints/data-solutions/vertex-mlops/metadata.yaml [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+spec:
+  title: MLOps with Vertex AI
+  source:
+    repo: https://github.com/GoogleCloudPlatform/cloud-foundation-fabric.git
```
Comment (juliocc) [inline blueprints/data-solutions/vertex-mlops/metadata.yaml:22]: did you figure out if there's a way to use a subdirectory?

**Review by juliocc — APPROVED** [review]

**Pull request merged**